### PR TITLE
WIP: Issue 4409: Add MultiThreadChecker implementation	

### PR DESCRIPTION
**Pull request by soon**
Issue #4409

WIP until #4869 (PR https://github.com/checkstyle/checkstyle/pull/4882) is merged. This PR also contains 2 commits - one from the #4882 and it will be removed as soon as #4882 is merged

This PR adds the MT mode for the Checker module

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java [hunk 1]
```diff
@@ -19,605 +19,13 @@
...
-    }
-
-    /**
-     * Sets the field haltOnException.
-     * @param haltOnException the new value.
-     */
-    public void setHaltOnException(boolean haltOnException) {
-        this.haltOnException = haltOnException;
-    }
+public class Checker extends AbstractChecker {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: Completely empty class is not acceptable.
If we override methods in MT Checker, than those methods should be abstract in AbstractChecker, and their implementation moved to non-MT Checker.

File: src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        }
+
+        return processFile(file, clones);
+    }
+
+    /**
+     * A check execution result.
+     * @author Andrew Kuchev.
+     */
+    private static final class ChecksExecutionResult {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java]: Expand documentation.
Checker calls filesets, not checks. So shouldn't this be called `CheckerExecutionResults`?

File: src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+     * @param file A file to process.
+     * @return A callable which processes the given file.
+     */
+    private Callable<ChecksExecutionResult> createProcessFileCallable(File file) {
+
+        final String fileName = file.getAbsolutePath();
+
+        return () -> {
+            ChecksExecutionResult result;
+            try {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java]: Shouldn't this section be a call to something in `AbstractChecker` as it's code looks similar?

File: src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+        /**
+         * Creates the completed execution result, i.e. with file and messages,
+         * but without thrown exception.
+         * @param file The processed file.
+         * @param localizedMessages The messages for the given file.
+         * @return The completed execution result.
+         */
+        public static ChecksExecutionResult getCompleted(File file,
+            SortedSet<LocalizedMessage> localizedMessages) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java]: Why do we recieve messages when we don't use them?

File: src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        /**
+         * Creates the failed execution result, i.e. with file and thrown exception,
+         * but without messages.
+         * @param file The processed file.
+         * @param thrownException The thrown exception.
+         * @return The failed execution result.
+         */
+        public static ChecksExecutionResult getFailed(File file,
+                CheckstyleException thrownException) {
+            return new ChecksExecutionResult(thrownException);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java]: Do we really need these methods for such simple actions?

Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java]: Removed

File: src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java (hunk 4 above)
Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java]: Fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java (hunk 3 above)
Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java]: Fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java (hunk 2 above)
Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java]: Renamed to `CheckerExecutionResults`.
Extended docs a bit. Previously this class also holds a file and file messages, but now it indicates only whether the execution successful or not. Probably, we could replace this class with just `CheckstyleException`

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java (hunk 1 above)
Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: Moved `processFilteredFiles` and `processFile` methods to `AbstractChecker` class.

Not sure about `process`, `finishFileSetChecksProcessing` and `destroyFileSetChecks` methods - all of them calls super method

**Comment by romani** [comment]
we released 8.2 version, please rebase all your PRs to our latest master to avoid CI problems

**Comment by shalugin** [comment]
@soon Can you rebase your PR?

**Comment by soon** [comment]
@shalugin I have rebased it to master, though I did not run tests on it and therefore current implementation might not work. I will fix these issues a bit later, thank you in advance for understanding.

**Comment by romani** [comment]
I am waiting for @rnveach  to confirm multi-file and multi-threading modes.

This PR unlikely to be merged, a lot need to be changed, but some part might be reused.

**Comment by daniilyar** [comment]
@soon , please rebase your code on the latest state of the master branch. Recently we added the config for Jenkins in master branch which is required for CI build to pass

**Comment by romani** [comment]
we cannot find agreement on design of MT mode, I am closing this PR, we can come back to it ones we agree and have time to continue.

**Pull request closed without merging**